parquet/src/main/java/org/apache/iceberg/data/parquet/GenericParquetReaders.java

```diff
@@ -53,21 +53,23 @@ private GenericParquetReaders() {
   }
 
   private static class RecordReader extends StructReader<Record, Record> {
-    private final StructType structType;
+    private final GenericRecord template;
 
     RecordReader(List<Type> types,
                  List<ParquetValueReader<?>> readers,
                  StructType struct) {
       super(types, readers);
-      this.structType = struct;
+      this.template = struct != null ? GenericRecord.create(struct) : null;
```

Comment:
Should these also have preconditions that validate the struct is not null?

Reply:
My original change did not contain the null check, and there were failing tests (org.apache.iceberg.TestGenericAppenderFactory > testPosDeleteWriterWithRowSchema[FileFormat=parquet, Partitioned=true] FAILED) with the exception below.

In nutshell, when we are reading the positional delete files we either have `row` in the data or we don't. I have tried to fix this by returning a `null` reader instead of creating a reader for a `null` struct. But this caused issues with Spark projection where we create the reader for the full schema, and having a `null` or missing reader caused issues when we tried to match the readers to the projected schema.
Fixing this is not trivial so I decided to move forward with the null checks to keep the backward compatibility:
- It should possible to create a reader with a `null` schema
- Fail only when we try to actually read from this reader

```
    java.lang.NullPointerException
        at org.apache.iceberg.data.GenericRecord.<init>(GenericRecord.java:63)
        at org.apache.iceberg.data.GenericRecord.create(GenericRecord.java:53)
        at org.apache.iceberg.data.parquet.GenericParquetReaders$RecordReader.<init>(GenericParquetReaders.java:62)
        at org.apache.iceberg.data.parquet.GenericParquetReaders.createStructReader(GenericParquetReaders.java:52)
        at org.apache.iceberg.data.parquet.BaseParquetReaders$ReadBuilder.struct(BaseParquetReaders.java:176)
        at org.apache.iceberg.data.parquet.BaseParquetReaders$ReadBuilder.struct(BaseParquetReaders.java:114)
        at org.apache.iceberg.parquet.TypeWithSchemaVisitor.visit(TypeWithSchemaVisitor.java:141)
        at org.apache.iceberg.parquet.TypeWithSchemaVisitor.visitField(TypeWithSchemaVisitor.java:169)
        at org.apache.iceberg.parquet.TypeWithSchemaVisitor.visitFields(TypeWithSchemaVisitor.java:183)
        at org.apache.iceberg.parquet.TypeWithSchemaVisitor.visit(TypeWithSchemaVisitor.java:47)
        at org.apache.iceberg.data.parquet.BaseParquetReaders.createReader(BaseParquetReaders.java:67)
        at org.apache.iceberg.data.parquet.BaseParquetReaders.createReader(BaseParquetReaders.java:58)
        at org.apache.iceberg.data.parquet.GenericParquetReaders.buildReader(GenericParquetReaders.java:41)
        at org.apache.iceberg.data.DeleteFilter.lambda$openDeletes$6(DeleteFilter.java:255)
        at org.apache.iceberg.parquet.ReadConf.<init>(ReadConf.java:118)
        at org.apache.iceberg.parquet.ParquetReader.init(ParquetReader.java:66)
        at org.apache.iceberg.parquet.ParquetReader.iterator(ParquetReader.java:77)
        at org.apache.iceberg.parquet.ParquetReader.iterator(ParquetReader.java:38)
        at org.apache.iceberg.util.Filter.lambda$filter$0(Filter.java:35)
        at org.apache.iceberg.io.CloseableIterable$2.iterator(CloseableIterable.java:73)
        at org.apache.iceberg.io.CloseableIterable$4$1.<init>(CloseableIterable.java:99)
        at org.apache.iceberg.io.CloseableIterable$4.iterator(CloseableIterable.java:98)
        at org.apache.iceberg.io.CloseableIterable$ConcatCloseableIterable$ConcatCloseableIterator.<init>(CloseableIterable.java:152)
        at org.apache.iceberg.io.CloseableIterable$ConcatCloseableIterable$ConcatCloseableIterator.<init>(CloseableIterable.java:143)
        at org.apache.iceberg.io.CloseableIterable$ConcatCloseableIterable.iterator(CloseableIterable.java:138)
        at org.apache.iceberg.io.CloseableIterable$ConcatCloseableIterable.iterator(CloseableIterable.java:129)
        at java.lang.Iterable.forEach(Iterable.java:74)
        at org.apache.iceberg.deletes.Deletes.toPositionIndex(Deletes.java:97)
        at org.apache.iceberg.deletes.Deletes.toPositionIndex(Deletes.java:91)
        at org.apache.iceberg.data.DeleteFilter.applyPosDeletes(DeleteFilter.java:231)
        at org.apache.iceberg.data.DeleteFilter.filter(DeleteFilter.java:126)
        at org.apache.iceberg.data.GenericReader.open(GenericReader.java:77)
        at org.apache.iceberg.relocated.com.google.common.collect.Iterators$6.transform(Iterators.java:826)
        at org.apache.iceberg.relocated.com.google.common.collect.TransformedIterator.next(TransformedIterator.java:52)
        at org.apache.iceberg.io.CloseableIterable$ConcatCloseableIterable$ConcatCloseableIterator.<init>(CloseableIterable.java:151)
        at org.apache.iceberg.io.CloseableIterable$ConcatCloseableIterable$ConcatCloseableIterator.<init>(CloseableIterable.java:143)
        at org.apache.iceberg.io.CloseableIterable$ConcatCloseableIterable.iterator(CloseableIterable.java:138)
        at org.apache.iceberg.data.GenericReader.open(GenericReader.java:65)
        at org.apache.iceberg.data.TableScanIterable.iterator(TableScanIterable.java:41)
        at org.apache.iceberg.data.TableScanIterable.iterator(TableScanIterable.java:29)
        at java.lang.Iterable.forEach(Iterable.java:74)
```